### ISSUE TITLE
ty 0.0.64 false-positives on lambdas passed to overloaded-callable Protocol union members (regression from 0.0.63)

### ISSUE THREAD

**Issue by fourtreestech** [astral-sh/ty#4098]
## Summary

`ty 0.0.64` reports a false-positive `invalid-argument-type` on any lambda
passed to a Reflex event-handler prop (e.g. `rx.select(on_change=lambda v: ...)`),
where `ty 0.0.63` and earlier accept the same code. Bisected across
0.0.55 / 0.0.60 / 0.0.62 / 0.0.63 (all pass) vs. 0.0.64 (fails).

This affects a real project: every lambda event handler in our Reflex app
(8 occurrences across 6 files) started failing `ty check` after upgrading from
0.0.63 to 0.0.64, with no corresponding code change. Notably this includes
Reflex's own documented pattern of passing a plain function reference
(`on_change=set_color_mode`, not even a lambda) to an event prop.

## Environment

- `ty 0.0.64 (5e64a131b 2026-07-27)`
- `reflex 0.9.6.post1`, `reflex-components-radix 0.9.5`
- Python 3.13
- macOS (Darwin 25.5.0)

## Reproduction

```python
# ty_repro.py — requires `reflex` installed (0.9.6.post1)
import reflex as rx


class S(rx.State):
    value: str = ""

    @rx.event
    def set_value(self, value: str):
        self.value = value


def component() -> rx.Component:
    return rx.select(
        ["a", "b"],
        on_change=lambda value: S.set_value(value),
    )
```

```
$ uvx ty@0.0.63 check ty_repro.py
All checks passed!

$ uvx ty@0.0.64 check ty_repro.py
error[invalid-argument-type]: Argument to function `Select.__call__` is incorrect
  --> ty_repro.py:28:9
   |
28 |         on_change=lambda value: S.set_value(value),
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Expected `EventType[()] | EventType[str] | None`, found `(value) -> EventCallback[()]`
   |
info: type `(value) -> EventCallback[()]` is not assignable to any element of the union `LambdaEventCallback[()] | EventCallback[()] | EventSpec | ... omitted 7 union elements`
info: ├── type `(value) -> EventCallback[()]` is not assignable to protocol `LambdaEventCallback[()]`
info: │   └── protocol member `__call__` is incompatible
info: ├── type `(value) -> EventCallback[()]` is not assignable to protocol `LambdaEventCallback[str]`
info: │   └── protocol member `__call__` is incompatible
info: └── ... omitted 8 union elements without additional context
info: Function defined here
   --> .venv/lib/python3.13/site-packages/reflex_components_radix/themes/components/select.pyi:855:9
    |
855 |     def __call__(
    |         ^^^^^^^^
    |
   ::: .venv/lib/python3.13/site-packages/reflex_components_radix/themes/components/select.pyi:952:9
    |
952 |         on_change: EventType[()] | EventType[str] | None = None,
    |         ------------------------------------------------------- Parameter declared here
    |

Found 1 diagnostic
```

## Expected behavior

`ty` accepts the lambda, as it did through 0.0.63, since it satisfies
`LambdaEventCallback[str].__call__(self, value: Var[str], /) -> Any` (Reflex's
`reflex_base/event/__init__.py`, `LambdaEventCallback` protocol at line ~2674):

```python
class LambdaEventCallback(Protocol[Unpack[P]]):
    __code__: types.CodeType

    @overload
    def __call__(self: "LambdaEventCallback[()]") -> Any: ...

    @overload
    def __call__(self: "LambdaEventCallback[V]", value: Var[V], /) -> Any: ...

    @overload
    def __call__(
        self: "LambdaEventCallback[V, V2]", value: Var[V], value2: Var[V2], /
    ) -> Any: ...
    ...
```

`on_change` is typed `EventType[str]`, i.e. (among other union members)
`LambdaEventCallback[str] | EventCallback[str] | ...`. A specialization to
`LambdaEventCallback[str]` should only expose the second overload
(`self: LambdaEventCallback[V]`) as applicable, which a 1-parameter lambda
satisfies. Instead 0.0.64 appears to check the lambda against every overload
regardless of the `self`-type discriminant, so the 0-arg and 2-arg overloads
also get checked and fail, sinking the whole union member.

## Notes

- I could not reduce this to a Reflex-free repro: a standalone `Protocol`
  with the same `self`-type-discriminated overloaded `__call__` shape, used
  via an equivalent `TypeAliasType` union parameter, type-checks fine on both
  0.0.63 and 0.0.64. So the trigger is more specific than "any protocol with
  self-discriminated overloads" — possibly interacting with `TypeVarTuple`/
  `Unpack`, the nested `TypeAliasType` layering in `reflex_base`'s `EventType`
  (`ItemOrList[LAMBDA_OR_STATE[...] | BASIC_EVENT_TYPES]`), or the `Generic`
  `EventCallback` descriptor class also present in the union. Happy to help
  narrow further if useful.
- The 0.0.64 changelog entry "Preserve receiver constraints when binding
  overloaded methods" looks like the most likely candidate change, but I
  haven't confirmed that by reverting just that patch.
- Reflex source for the types referenced above:
  `reflex_base/event/__init__.py` in the `reflex-base` package (pulled in
  transitively by `reflex>=0.9.6`).

**Comment by charliermarsh** [comment]
Codex gave me this reproduction:
```python
from typing import Any, Protocol, TypeVar, TypeVarTuple, Unpack, overload

P = TypeVarTuple("P")
V = TypeVar("V")


class Callback(Protocol[Unpack[P]]):
    @overload
    def __call__(self: "Callback[()]") -> Any: ...

    @overload
    def __call__(self: "Callback[V]", value: V, /) -> Any: ...

    def __call__(self, *args: Any) -> Any: ...


def takes_callback(callback: Callback[str]) -> None: ...


takes_callback(lambda value: value)
```

**Comment by charliermarsh** [comment]
Allegedly, this is from https://github.com/astral-sh/ruff/pull/27083. Previously, Reflex's generic `TypeAliasType` resolved to `Unknown`, so we weren't checking it at all. We need to fix the overload selection here so that the first `__call__` is omitted.
